pkg/proxy/usergating/handler.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package usergating
+
+import (
+	"context"
+	"net/http"
+
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog/v2"
+
+	"github.com/redhat-cps/front-proxy/pkg/proxy/usergating/ruleset"
+)
+
+type UserGate struct {
+	GetRuleSet ruleset.RuleSetGetter
+}
+
+func WithUserGating(delegate http.Handler, failed http.Handler, getRuleSet ruleset.RuleSetGetter) http.Handler {
+	gate := &UserGate{
+		GetRuleSet: getRuleSet,
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		u, ok := request.UserFrom(ctx)
+		if !ok {
+			delegate.ServeHTTP(w, r)
+			return
+		}
+
+		decision := gate.Authorize(ctx, u)
+		if decision != authorizer.DecisionDeny {
```

Comment:
Do we want to allow NoOpinion to be allowed in? cc @s-urbaniak

Reply:
This is an interesting point, although it's not making the actual authz decision (that will still happen at the shard) - we do have an allowed and denied list, @csams what's the intended behavior if a user is neither explicitly allowed or denied via the ruleset?

It seems like we should perhaps only have `allowed_user_names` then deny all other users, or else this isn't really a gate to prevent access by undefined random users?